storage/innobase/os/os0file.cc

```diff
@@ -4771,31 +4790,100 @@ readahead requests. */
 
 #if defined(LINUX_NATIVE_AIO)
 /*******************************************************************//**
+Submit buffered AIO requests on the given segment to the kernel. */
+UNIV_INTERN
+void
+os_aio_linux_dispatch_read_array_submit()
+{
+	os_aio_array_t* array = os_aio_read_array;
+	ulint total_submitted = 0;
+	if (!srv_use_native_aio) {
+		return;
+	}
+	os_mutex_enter(array->mutex);
+	/* Submit aio requests buffered on all segments. */
+	for (ulint i = 0; i < array->n_segments; i++) {
+		int count = array->count[i];
+		if (count > 0) {
+			ulint iocb_index = i * array->n_slots
+					   / array->n_segments;
+			int submitted;
+			submitted = io_submit(array->aio_ctx[i], count,
+					      &(array->pending[iocb_index]));
+			if (submitted == count) {
+				total_submitted += submitted;
+			} else {
+				/* io_submit returns number of successfully
+				queued requests or -errno. */
+				fprintf(stderr, "Trying to sumbit %d aio "
+					"requests, io_submit returns %d.\n",
+					count, submitted);
+				if (submitted < 0) {
+					errno = -submitted;
```

Comment:
1) Is this the actual libaio contract?
2) If it is and we get the right errno value here, what consumes it?

Reply:
1. Yes, it is true  for libaio. See 'NOTES' section in http://man7.org/linux/man-pages/man2/io_submit.2.html
2. Changed so that `errno` is printed to `ib_log`